Panic on Route53 records management

**What happened**:
We are managing our AWS route53 DNS records with externalDNS.
For a very specific purpose, we had to manually change a previously external DNS managed record.
Then, the externalDNS container started to crash and keep restarting time to time with the following logs: 

```
time="2025-07-10T14:16:05Z" level=debug msg="Refreshing zones list cache"
time="2025-07-10T14:16:05Z" level=debug msg="Considering zone: /hostedzone/REDACTED (domain: REDACTED.)"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Endpoints generated from ingress: REDACTED: [REDACTED 0 IN CNAME  REDACTED [] REDACTED 0 IN CNAME  REDACTED []]"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Removing duplicate endpoint REDACTED 0 IN CNAME  REDACTED []"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Modifying endpoint: REDACTED 0 IN CNAME  REDACTED [], setting alias=true"
time="2025-07-10T14:16:06Z" level=debug msg="Refreshing zones list cache"
time="2025-07-10T14:16:06Z" level=debug msg="Considering zone: /hostedzone/REDACTED (domain: REDACTED.)"
time="2025-07-10T14:16:06Z" level=info msg="Applying provider record filter for domains: [REDACTED. .REDACTED.]"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="ignoring record REDACTED that does not match domain filter"
time="2025-07-10T14:16:06Z" level=debug msg="Couldn't parse REDACTED as an IP address: ParseAddr(\"REDACTED\"): unexpected character (at \"REDACTED\")" comparisonTargets=REDACTED targets=REDACTED
time="2025-07-10T14:16:06Z" level=debug msg="Couldn't parse REDACTED as an IP address: ParseAddr(\"REDACTED\"): unexpected character (at \"REDACTED\")" comparisonTargets=REDACTED targets=REDACTED
time="2025-07-10T14:16:06Z" level=debug msg="Couldn't parse REDACTED as an IP address: ParseAddr(\"REDACTED\"): unexpected character (at \"REDACTED\")" comparisonTargets=REDACTED targets=REDACTED
time="2025-07-10T14:16:06Z" level=debug msg="Couldn't parse REDACTED as an IP address: ParseAddr(\"REDACTED\"): unexpected character (at \"REDACTED\")" comparisonTargets=REDACTED targets=REDACTED
time="2025-07-10T14:16:06Z" level=debug msg="Couldn't parse REDACTED as an IP address: ParseAddr(\"REDACTED\"): unexpected character (at \"REDACTED\")" comparisonTargets=REDACTED targets=REDACTED
time="2025-07-10T14:16:06Z" level=debug msg="Couldn't parse REDACTED as an IP address: ParseAddr(\"REDACTED\"): unexpected character (at \"REDACTED\")" comparisonTargets=REDACTED targets=REDACTED
time="2025-07-10T14:16:06Z" level=debug msg="Skipping endpoint REDACTED 300 IN AAAA  REDACTED [{aws/evaluate-target-health true} {alias true}] because owner id does not match for one or more items to create, found: \"\", required: \"public\""
time="2025-07-10T14:16:06Z" level=debug msg="Skipping endpoint REDACTED 300 IN AAAA  REDACTED [{aws/evaluate-target-health true} {alias true}] because owner id does not match, found: \"\", required: \"public\""
time="2025-07-10T14:16:06Z" level=debug msg="Skipping endpoint REDACTED 300 IN AAAA  REDACTED [{aws/evaluate-target-health true} {alias true}] because owner id does not match, found: \"\", required: \"public\""
time="2025-07-10T14:16:06Z" level=debug msg="Skipping endpoint REDACTED 0 IN AAAA  REDACTED [{alias true} {aws/evaluate-target-health true}] because owner id does not match, found: \"\", required: \"public\""
time="2025-07-10T14:16:06Z" level=debug msg="Refreshing zones list cache"
time="2025-07-10T14:16:06Z" level=debug msg="Considering zone: /hostedzone/REDACTED (domain: REDACTED.)"
panic: runtime error: index out of range [3] with length 3

goroutine 1 [running]:
sigs.k8s.io/external-dns/provider/aws.(*AWSProvider).createUpdateChanges(0xc00096fbc0, {0xc000497680, 0x5, 0xd?}, {0xc0009b9698, 0x3, 0x40?})
sigs.k8s.io/external-dns/provider/aws/aws.go:603 +0x5bc
sigs.k8s.io/external-dns/provider/aws.(*AWSProvider).ApplyChanges(0xc00096fbc0, {0x46723a0, 0xc000996750}, 0xc0004f3200)
sigs.k8s.io/external-dns/provider/aws/aws.go:642 +0xc5
sigs.k8s.io/external-dns/registry.(*TXTRegistry).ApplyChanges(0xc0009d8680, {0x46723a0, 0xc000996750}, 0xc0004f29c0)
sigs.k8s.io/external-dns/registry/txt.go:324 +0x9c8
sigs.k8s.io/external-dns/controller.(*Controller).RunOnce(0xc00096fc80, {0x46723d8, 0xc0003cbae0})
sigs.k8s.io/external-dns/controller/controller.go:258 +0x6e8
sigs.k8s.io/external-dns/controller.(*Controller).Run(0xc00096fc80, {0x46723d8, 0xc0003cbae0})
sigs.k8s.io/external-dns/controller/controller.go:361 +0xc5
main.main()
sigs.k8s.io/external-dns/main.go:451 +0x3b4f
```

**What you expected to happen**:

ExternalDNS to not crash in panic.

**How to reproduce it (as minimally and precisely as possible)**:
We have specific DNS records that I am not able to share for privacy purposes unfortunately.

**Anything else we need to know?**:
When we changed the DNS records manually, we didn't clean the TXT records which are used by externalDNS for ownership purposes.
This might help you to diagnose.

**Environment**: 
- External-DNS version: v0.16.1
- DNS provider: AWS route53

--- a/provider/aws/aws.go
+++ b/provider/aws/aws.go
@@ -637,6 +637,10 @@ func (p *AWSProvider) createUpdateChanges(newEndpoints, oldEndpoints []*endpoint
 	var updates []*endpoint.Endpoint
 
 	for i, newE := range newEndpoints {
+		if i >= len(oldEndpoints) || oldEndpoints[i] == nil {
+			log.Debugf("skip %s as endpoint not found in current endpoints", newE.DNSName)
+			continue
+		}
 		oldE := oldEndpoints[i]
 		if p.requiresDeleteCreate(oldE, newE) {
 			deletes = append(deletes, oldE)

--- a/provider/aws/aws.go
+++ b/provider/aws/aws.go
@@ -637,11 +637,11 @@ func (p *AWSProvider) createUpdateChanges(newEndpoints, oldEndpoints []*endpoint
 	var updates []*endpoint.Endpoint
 
 	for i, newE := range newEndpoints {
-		oldE := oldEndpoints[i]
-		if i >= len(oldEndpoints) || oldE == nil {
+		if i >= len(oldEndpoints) || oldEndpoints[i] == nil {
 			log.Debugf("skip %s as endpoint not found in current endpoints", newE.DNSName)
 			continue
 		}
+		oldE := oldEndpoints[i]
 		if p.requiresDeleteCreate(oldE, newE) {
 			deletes = append(deletes, oldE)
 			creates = append(creates, newE)

--- a/provider/aws/aws.go
+++ b/provider/aws/aws.go
@@ -637,11 +637,11 @@ func (p *AWSProvider) createUpdateChanges(newEndpoints, oldEndpoints []*endpoint
 	var updates []*endpoint.Endpoint
 
 	for i, newE := range newEndpoints {
-		if i >= len(oldEndpoints) || oldEndpoints[i] == nil {
+		oldE := oldEndpoints[i]
+		if i >= len(oldEndpoints) || oldE == nil {
 			log.Debugf("skip %s as endpoint not found in current endpoints", newE.DNSName)
 			continue
 		}
-		oldE := oldEndpoints[i]
 		if p.requiresDeleteCreate(oldE, newE) {
 			deletes = append(deletes, oldE)
 			creates = append(creates, newE)

--- a/provider/aws/aws_test.go
+++ b/provider/aws/aws_test.go
@@ -2817,3 +2817,35 @@ func TestGeoProximityWithBias(t *testing.T) {
 		})
 	}
 }
+
+func TestAWSProvider_createUpdateChanges_NewMoreThanOld(t *testing.T) {
+	provider, _ := newAWSProvider(t, endpoint.NewDomainFilter([]string{"foo.bar."}), provider.NewZoneIDFilter([]string{}), provider.NewZoneTypeFilter(""), true, false, nil)
+
+	oldEndpoints := []*endpoint.Endpoint{
+		endpoint.NewEndpointWithTTL("record1.foo.bar.", endpoint.RecordTypeA, endpoint.TTL(300), "1.1.1.1"),
+		nil,
+	}
+	newEndpoints := []*endpoint.Endpoint{
+		endpoint.NewEndpointWithTTL("record1.foo.bar.", endpoint.RecordTypeA, endpoint.TTL(300), "1.1.1.1"),
+		endpoint.NewEndpointWithTTL("record2.foo.bar.", endpoint.RecordTypeA, endpoint.TTL(300), "2.2.2.2"),
+	}
+
+	changes := provider.createUpdateChanges(newEndpoints, oldEndpoints)
+
+	// record2 should be created, record1 should be upserted
+	var creates, upserts, deletes int
+	for _, c := range changes {
+		switch c.Action {
+		case route53types.ChangeActionCreate:
+			creates++
+		case route53types.ChangeActionUpsert:
+			upserts++
+		case route53types.ChangeActionDelete:
+			deletes++
+		}
+	}
+
+	require.Equal(t, 0, creates, "should create the extra new endpoint")
+	require.Equal(t, 1, upserts, "should upsert the matching endpoint")
+	require.Equal(t, 0, deletes, "should not delete anything")
+}

--- a/provider/aws/aws_test.go
+++ b/provider/aws/aws_test.go
@@ -2828,6 +2828,7 @@ func TestAWSProvider_createUpdateChanges_NewMoreThanOld(t *testing.T) {
 	newEndpoints := []*endpoint.Endpoint{
 		endpoint.NewEndpointWithTTL("record1.foo.bar.", endpoint.RecordTypeA, endpoint.TTL(300), "1.1.1.1"),
 		endpoint.NewEndpointWithTTL("record2.foo.bar.", endpoint.RecordTypeA, endpoint.TTL(300), "2.2.2.2"),
+		endpoint.NewEndpointWithTTL("record3.foo.bar.", endpoint.RecordTypeA, endpoint.TTL(300), "3.3.3.3"),
 	}
 
 	changes := provider.createUpdateChanges(newEndpoints, oldEndpoints)

